CRM/Civicase/Service/CaseTypeCategoryEventHandler.php

```diff
@@ -0,0 +1,236 @@
+    $this->customFieldExtends = $customFieldExtends;
+  }
+
+  /**
+   * Perform actions on case type category create.
+   *
+   * @param string $name
+   *   Case type category name.
+   */
+  public function onCreate($name) {
+    if (!$name || !$this->shouldRun()) {
+      return;
+    }
+
+    $this->menu->createItems($name);
+    $this->customFieldExtends->create($name, "Case ({$name})");
+    $this->customData->create($name);
+  }
+
+  /**
+   * Perform actions on case type category create.
+   *
+   * @param array $category
+   *   Case type category data, should contain:
+   *   id - category id,
+   *   is_active - is_active category property (enabled/disabled),
+   *   icon - category icon.
+   */
+  public function onUpdate(array $category) {
+    if (empty($category['id']) || !$this->shouldRun()) {
+      return;
+    }
+
+    $updateParams = [];
+    if (isset($category['is_active'])) {
+      $updateParams['is_active'] = !empty($category['is_active']) ? 1 : 0;
+    }
+    if (isset($category['icon'])) {
+      $updateParams['icon'] = 'crm-i ' . $category['icon'];
+    }
+
+    if (!empty($updateParams)) {
+      $this->menu->updateItems($category['id'], $updateParams);
+    }
+  }
+
+  /**
+   * Removes case type category menu item from the civicrm navigation bar.
+   *
+   * @param string $name
+   *   Case type category name.
+   */
+  public function onDelete($name) {
+    if (!$name || !$this->shouldRun()) {
+      return;
+    }
+
+    $this->menu->deleteItems($name);
+    $this->customFieldExtends->delete($name);
+    $this->customData->delete($name);
+  }
+
+  /**
+   * Set items to check before running action (onCreate, onUpdate, onDelete).
+   *
+   * You may use this method to override default check settings and provide
+   * necessary data.
+   *
+   * @param array $params
+   *   Parameters that are used to perform necessary checks.
+   */
+  public function setCheck(array $params) {
+    if (isset($params['needCheckReferer'])) {
+      $this->needCheckReferer = $params['needCheckReferer'] ? TRUE : FALSE;
+    }
+    if (isset($params['needCheckOptionGroup'])) {
+      $this->needCheckOptionGroup = $params['needCheckOptionGroup'] ? TRUE : FALSE;
+    }
+    if (isset($params['optionGroupId'])) {
+      $this->optionGroupId = intval($params['optionGroupId']);
+    }
+    if (isset($params['objectName'])) {
+      $this->objectName = $params['objectName'];
+    }
+  }
+
+  /**
+   * Determines if the hook will run.
+   *
+   * Keep in mind that referer will be checked if $needCheckReferer property
+   * is set to TRUE.
+   * Keep in mind that option group will be checked if $needCheckOptionGroup
+   * is TRUE and $optionGroupId is set.
+   * By default this method would always return TRUE.
+   *
+   * @see setCheck()
+   *
+   * @return bool
+   *   TRUE if passed object is an option value of Case Type Categories
+   *   option group and http referer is correct, FALSE otherwise.
+   */
+  protected function shouldRun() {
```

Comment:
We don't need a should run function in this class. The check referrer function is not needed here also. The check referrer function should be in the listener class and that check should be made in the listener class before this service is called. This class should basically handle the events, checks peculiar to the calling class should be made there.

Reply:
In case if we need this somewhere else we have to duplicate code from one listener/hook processing to another. Why not to have it one place?

Reply:
The check for http referrer is needed in the listener class ONLY. It is the responsibility of that class to check if the request is coming from the option value page before utilizing the handler class. 
Checks related to optiob value name and ID is for the post process hook for the option value form ONLY. It is the responsibility of that class to check if the right option value is being created/edited/updated.

Bear in mind that this class is an event handler class that only utilizes a set of services. If there needs to be checks, the checks should be made in the individual services themselves. For example it is the responsibility of the Case category menu class to check if the case category passed is valid or not.